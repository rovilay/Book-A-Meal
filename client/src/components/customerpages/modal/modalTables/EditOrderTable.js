/* eslint jsx-a11y/label-has-for:0 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ReactPaginate from 'react-paginate';
import classname from 'classnames';

import { editOrderTableHead } from '../../../../helpers/tableHeadData';
import CustomerOrderTableRow from '../../orderPage/CustomerOrderTablerow';

class EditOrderTable extends Component {
  constructor(props) {
    super(props);

    this.handlePaginationClick = this.handlePaginationClick.bind(this);
    this.submitEdit = this.submitEdit.bind(this);
  }

  /**
   * handles pagination changes
   *
   * @memberof EditOrderTable
   * @param {object} data data object from pagination component
   */
  handlePaginationClick(data) {
    const {
      limit,
    } = this.props.orderedMealsPagination;

    const {
      UserId,
      orderId
    } = this.props.editOrder;

    const nextPage = data.selected;
    const newOffset = nextPage * limit;
    const mealsUrl = `/api/v1/orders/${UserId}?${orderId}`;
    this.props.getOrderMeals(mealsUrl, { limit, offset: newOffset });
  }

  /**
   * Submits the edited order
   * @param  {any} event DOM event
   * @return {void}
   * @memberof EditOrderTable
   */
  submitEdit(event) {
    event.preventDefault();
    const {
      updateOrder,
      editOrder
    } = this.props;

    const deliveryAddress = document.getElementById('delivery-address').value.trim();
    const data = { deliveryAddress, meals: editOrder.orderedMeals };
    updateOrder(editOrder.orderId, data);
  }

  render() {
    const {
      title: modalTitle,
      editOrder,
      orderedMealsPagination
    } = this.props;

    const {
      orderedMeals,
      totalPrice,
    } = editOrder;

    const {
      offset,
      numOfPages,
      count
    } = orderedMealsPagination;

    return (
      <div className="modal-container">
        <h2 className="title">
          {modalTitle}
        </h2>

        <hr />

        <form
          onSubmit={this.submitEdit}
          className="edit-order"
        >
          <div className="address">
            <label htmlFor="address">
              Address:
            </label>
            <div className="input-div">
              <input
                type="text"
                placeholder="Enter delivery address"
                name="deliveryAddress"
                id="delivery-address"
                defaultValue={editOrder.deliveryAddress}
                required
              />
            </div>
          </div>

          <div className="container-test">
            <div className="row head">
              {
                editOrderTableHead.map((head, i) => (
                  <p
                    key={i}
                    className={classname('row-item', { 'meal-title': head === 'Meal' || head === 'orderId' })}
                  >
                    {head}
                  </p>
                ))
              }
            </div>
            {
              orderedMeals.map((meal, i) => {
                const {
                  id,
                  title: Meal,
                  cost,
                  portion,
                  price
                } = meal;
                const item = {
                  sn: ++i + offset,
                  Meal,
                  cost,
                  portion,
                  price
                };

                return (
                  <CustomerOrderTableRow
                    key={id}
                    item={item}
                    mealId={id}
                    {...this.props}
                    orderedMealsLength={orderedMeals.length}
                    actions={{
                      edit: false,
                      info: false,
                      delete: true
                    }}
                  />
                );
              })
            }
          </div>
          <div className="order">
            <p id="order-total-price" className="orderTot">
              Total Price: &#8358;{totalPrice}
            </p>
            <button
              type="submit"
              name="orderbtn"
              id="order-btn"
              className="btn-2"
              disabled={orderedMeals.length <= 0}
            >
              Update order
            </button>
          </div>
        </form>
        {
          (count > 5)
          &&
          <div className="modal-pagination">
            <ReactPaginate
              previousLabel="<<"
              nextLabel=">>"
              breakLabel={<a href="">...</a>}
              breakClassName="break-me"
              pageCount={numOfPages}
              marginPagesDisplayed={2}
              pageRangeDisplayed={5}
              onPageChange={this.handlePaginationClick}
              containerClassName="pagination"
              subContainerClassName="pages pagination"
              activeClassName="active"
            />
          </div>
        }
      </div>
    );
  }
}

EditOrderTable.propTypes = {
  title: PropTypes.string.isRequired,
  editOrder: PropTypes.object.isRequired,
  updateOrder: PropTypes.func.isRequired,
  getOrderMeals: PropTypes.func.isRequired,
  orderedMealsPagination: PropTypes.object.isRequired
};

export default EditOrderTable;
