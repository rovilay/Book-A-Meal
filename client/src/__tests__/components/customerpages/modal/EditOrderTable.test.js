import React from 'react';
import { shallow } from 'enzyme';
import { orderOnEdit } from '../../../__mockData__/orderMock';
import EditOrderTable from '../../../../components/customerpages/modal/modalTables/EditOrderTable';

describe('EditOrderTable component test', () => {
  const setup = () => {
    const props = {
      orderedMealsPagination: {
        limit: 12,
        count: 13,
        numOfPages: 2
      },
      getOrderMeals: () => jest.fn(),
      updateOrder: () => jest.fn(),
      title: 'Edit Order',
      editOrder: {
        ...orderOnEdit,
        UserId: '1'
      }
    };

    return shallow(<EditOrderTable {...props} />)
  };


  it('should render EditOrderTable correctly', (done) => {
    const wrapper = setup();

    expect(wrapper).toMatchSnapshot();

    done();
  });


  it('should call `handlePaginationClick` if pagination button is clicked',
  (done) => {
    const wrapper = setup();

    const data = {
      selected: 1
    };

    const handlePaginationClickSpy= jest.spyOn(
      wrapper.instance(), 'handlePaginationClick'
    );
    wrapper.instance().handlePaginationClick(data)
    expect(handlePaginationClickSpy).toHaveBeenCalled();

    done();
  });


  it('should call `submitEdit` if update order button is clicked',
  (done) => {
    const wrapper = setup();

    const event = {
      preventDefault: () => jest.fn()
    }

    document.body.innerHTML =
    '<div>' +
    '  <input type="text" id="delivery-address" value="isuti rd" />'
    '</div>';

    const submitEditSpy= jest.spyOn(wrapper.instance(), 'submitEdit');
    wrapper.instance().submitEdit(event)
    expect(submitEditSpy).toHaveBeenCalled();

    done();
  });
});
