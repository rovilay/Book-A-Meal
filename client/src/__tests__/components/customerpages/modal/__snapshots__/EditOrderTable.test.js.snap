// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditOrderTable component test should render EditOrderTable correctly 1`] = `
<div
  className="modal-container"
>
  <h2
    className="title"
  >
    Edit Order
  </h2>
  <hr />
  <form
    className="edit-order"
    onSubmit={[Function]}
  >
    <div
      className="address"
    >
      <label
        htmlFor="address"
      >
        Address:
      </label>
      <div
        className="input-div"
      >
        <input
          defaultValue="isuti rd"
          id="delivery-address"
          name="deliveryAddress"
          placeholder="Enter delivery address"
          required={true}
          type="text"
        />
      </div>
    </div>
    <div
      className="container-test"
    >
      <div
        className="row head"
      >
        <p
          className="row-item"
          key="0"
        >
          S/N
        </p>
        <p
          className="row-item meal-title"
          key="1"
        >
          Meal
        </p>
        <p
          className="row-item"
          key="2"
        >
          Unit Cost
        </p>
        <p
          className="row-item"
          key="3"
        >
          Portion
        </p>
        <p
          className="row-item"
          key="4"
        >
          Cost
        </p>
        <p
          className="row-item"
          key="5"
        >
          Actions
        </p>
      </div>
      <CustomerOrderTableRow
        actions={
          Object {
            "delete": true,
            "edit": false,
            "info": false,
          }
        }
        editOrder={
          Object {
            "UserId": "1",
            "deliveryAddress": "isuti rd",
            "orderId": 5,
            "orderedMeals": Array [
              Object {
                "cost": 300,
                "id": "1",
                "portion": 1,
                "price": 300,
                "title": "EBA AND EGUSI",
              },
              Object {
                "cost": 200,
                "id": "2",
                "portion": 1,
                "price": 200,
                "title": "SALAD",
              },
              Object {
                "cost": 100,
                "id": "3",
                "portion": 2,
                "price": 200,
                "title": "SPAGHETTI",
              },
              Object {
                "cost": 400,
                "id": "4",
                "portion": 1,
                "price": 400,
                "title": "YAM AND EGG",
              },
            ],
            "totalPrice": 1100,
          }
        }
        getOrderMeals={[Function]}
        isEdit={false}
        item={
          Object {
            "Meal": "EBA AND EGUSI",
            "cost": 300,
            "portion": 1,
            "price": 300,
            "sn": NaN,
          }
        }
        key="1"
        mealId="1"
        orderCreatedAt={"2018-09-05T01:49:08.594Z"}
        orderedMealsLength={4}
        orderedMealsPagination={
          Object {
            "count": 13,
            "limit": 12,
            "numOfPages": 2,
          }
        }
        title="Edit Order"
        updateOrder={[Function]}
      />
      <CustomerOrderTableRow
        actions={
          Object {
            "delete": true,
            "edit": false,
            "info": false,
          }
        }
        editOrder={
          Object {
            "UserId": "1",
            "deliveryAddress": "isuti rd",
            "orderId": 5,
            "orderedMeals": Array [
              Object {
                "cost": 300,
                "id": "1",
                "portion": 1,
                "price": 300,
                "title": "EBA AND EGUSI",
              },
              Object {
                "cost": 200,
                "id": "2",
                "portion": 1,
                "price": 200,
                "title": "SALAD",
              },
              Object {
                "cost": 100,
                "id": "3",
                "portion": 2,
                "price": 200,
                "title": "SPAGHETTI",
              },
              Object {
                "cost": 400,
                "id": "4",
                "portion": 1,
                "price": 400,
                "title": "YAM AND EGG",
              },
            ],
            "totalPrice": 1100,
          }
        }
        getOrderMeals={[Function]}
        isEdit={false}
        item={
          Object {
            "Meal": "SALAD",
            "cost": 200,
            "portion": 1,
            "price": 200,
            "sn": NaN,
          }
        }
        key="2"
        mealId="2"
        orderCreatedAt={"2018-09-05T01:49:08.594Z"}
        orderedMealsLength={4}
        orderedMealsPagination={
          Object {
            "count": 13,
            "limit": 12,
            "numOfPages": 2,
          }
        }
        title="Edit Order"
        updateOrder={[Function]}
      />
      <CustomerOrderTableRow
        actions={
          Object {
            "delete": true,
            "edit": false,
            "info": false,
          }
        }
        editOrder={
          Object {
            "UserId": "1",
            "deliveryAddress": "isuti rd",
            "orderId": 5,
            "orderedMeals": Array [
              Object {
                "cost": 300,
                "id": "1",
                "portion": 1,
                "price": 300,
                "title": "EBA AND EGUSI",
              },
              Object {
                "cost": 200,
                "id": "2",
                "portion": 1,
                "price": 200,
                "title": "SALAD",
              },
              Object {
                "cost": 100,
                "id": "3",
                "portion": 2,
                "price": 200,
                "title": "SPAGHETTI",
              },
              Object {
                "cost": 400,
                "id": "4",
                "portion": 1,
                "price": 400,
                "title": "YAM AND EGG",
              },
            ],
            "totalPrice": 1100,
          }
        }
        getOrderMeals={[Function]}
        isEdit={false}
        item={
          Object {
            "Meal": "SPAGHETTI",
            "cost": 100,
            "portion": 2,
            "price": 200,
            "sn": NaN,
          }
        }
        key="3"
        mealId="3"
        orderCreatedAt={"2018-09-05T01:49:08.594Z"}
        orderedMealsLength={4}
        orderedMealsPagination={
          Object {
            "count": 13,
            "limit": 12,
            "numOfPages": 2,
          }
        }
        title="Edit Order"
        updateOrder={[Function]}
      />
      <CustomerOrderTableRow
        actions={
          Object {
            "delete": true,
            "edit": false,
            "info": false,
          }
        }
        editOrder={
          Object {
            "UserId": "1",
            "deliveryAddress": "isuti rd",
            "orderId": 5,
            "orderedMeals": Array [
              Object {
                "cost": 300,
                "id": "1",
                "portion": 1,
                "price": 300,
                "title": "EBA AND EGUSI",
              },
              Object {
                "cost": 200,
                "id": "2",
                "portion": 1,
                "price": 200,
                "title": "SALAD",
              },
              Object {
                "cost": 100,
                "id": "3",
                "portion": 2,
                "price": 200,
                "title": "SPAGHETTI",
              },
              Object {
                "cost": 400,
                "id": "4",
                "portion": 1,
                "price": 400,
                "title": "YAM AND EGG",
              },
            ],
            "totalPrice": 1100,
          }
        }
        getOrderMeals={[Function]}
        isEdit={false}
        item={
          Object {
            "Meal": "YAM AND EGG",
            "cost": 400,
            "portion": 1,
            "price": 400,
            "sn": NaN,
          }
        }
        key="4"
        mealId="4"
        orderCreatedAt={"2018-09-05T01:49:08.594Z"}
        orderedMealsLength={4}
        orderedMealsPagination={
          Object {
            "count": 13,
            "limit": 12,
            "numOfPages": 2,
          }
        }
        title="Edit Order"
        updateOrder={[Function]}
      />
    </div>
    <div
      className="order"
    >
      <p
        className="orderTot"
        id="order-total-price"
      >
        Total Price: â‚¦
        1100
      </p>
      <button
        className="btn-2"
        disabled={false}
        id="order-btn"
        name="orderbtn"
        type="submit"
      >
        Update order
      </button>
    </div>
  </form>
  <div
    className="modal-pagination"
  >
    <PaginationBoxView
      activeClassName="active"
      breakClassName="break-me"
      breakLabel={
        <a
          href=""
        >
          ...
        </a>
      }
      containerClassName="pagination"
      disableInitialCallback={false}
      disabledClassName="disabled"
      marginPagesDisplayed={2}
      nextClassName="next"
      nextLabel=">>"
      onPageChange={[Function]}
      pageCount={2}
      pageRangeDisplayed={5}
      previousClassName="previous"
      previousLabel="<<"
      subContainerClassName="pages pagination"
    />
  </div>
</div>
`;
